File: slash/components/trivia.ts
import { ButtonStyle } from "discord-api-types/v10";
import { ActionRow, Button, Interaction } from "discord-workers";

export default async (
  int: Interaction,
  wait: (f: any) => void
): Promise<Response | void> => {
  console.log(JSON.stringify(int));
  if (int.message?.interaction?.name === "trivia") {
    const options = new ActionRow();
    for (const choice of int.message.components![0].components as any[]) {
      options.component(
        new Button()
          .customID(choice.custom_id)
          .label(choice.label)
          .disabled()
          .style(
            (int.data as any).custom_id === "correct"
              ? choice.custom_id === "correct"
                ? ButtonStyle.Success
                : ButtonStyle.Secondary
              : (int.data as any).custom_id === choice.custom_id
              ? ButtonStyle.Danger
              : choice.custom_id === "correct"
              ? ButtonStyle.Success
              : ButtonStyle.Secondary
          )
      );
    }

    wait(
      int.edit({
        content: int.message?.content,
        components: [options.toJSON()],
      })
    );

    return int.respond(6, {});
  }
};

File: backend/components/trivia.go
package components

import (
	"backend/types"
	"encoding/json"
	"fmt"
)

func trivia(i *types.Interaction, data *types.MessageComponentInteractionData) {
	options := types.ActionRow{
		Type:       types.ComponentActionRow,
		Components: []interface{}{},
	}

	for _, choice := range i.Message.Components[0].Components {
		encoded, _ := json.Marshal(choice)
		var btn types.Button
		_ = json.Unmarshal(encoded, &btn)

		var style types.ButtonStyle
		if data.CustomID == "correct" {
			if btn.CustomID == "correct" {
				style = types.ButtonSuccess
			} else {
				style = types.ButtonSecondary
			}
		} else if data.CustomID == btn.CustomID {
			style = types.ButtonDanger
		} else if btn.CustomID == "correct" {
			style = types.ButtonSuccess
		} else {
			style = types.ButtonSecondary
		}

		options.Components = append(options.Components, types.Button{
			Type:     types.ComponentButton,
			Label:    btn.Label,
			Disabled: true,
			CustomID: btn.CustomID,
			Style:    style,
		})
	}

	err := i.Respond(types.InteractionResponse{
		Type: types.ResponseUpdateMessage,
		Data: types.ResponseData{
			Content:    i.Message.Content,
			Components: []interface{}{options},
		},
	})

	if err != nil {
		fmt.Println(err)
	}
}
